rest/api.go

```diff
@@ -0,0 +1,243 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package rest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus-community/promql-langserver/internal/vendored/go-tools/lsp/protocol"
+	"github.com/prometheus-community/promql-langserver/langserver"
+	promClient "github.com/prometheus-community/promql-langserver/prometheus"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/route"
+)
+
+func returnJSON(w http.ResponseWriter, content interface{}) {
+	encoder := json.NewEncoder(w)
+
+	err := encoder.Encode(content)
+	if err != nil {
+		http.Error(w, errors.Wrapf(err, "failed to write response").Error(), 500)
+	}
+}
+
+type lspData struct {
+	// Expr is the promQL expression and is required for all endpoint available.
+	Expr string `json:"expr"`
+	// Limit is the number max of result returned to the client. It will be used for the autocompletion and the diagnostics.
+	Limit *uint64 `json:"limit,omitempty"`
+	// PositionLine is the number of the line for which the metadata is queried.
+	PositionLine *float64 `json:"positionLine,omitempty"`
+	// PositionChar for which the metadata is queried. Characters are counted as UTF16 Codepoints.
+	PositionChar *float64 `json:"positionChar,omitempty"`
+}
+
+func (d *lspData) UnmarshalJSON(data []byte) error {
+	var tmp lspData
+	type plain lspData
+	if err := json.Unmarshal(data, (*plain)(&tmp)); err != nil {
+		return err
+	}
+	if len(tmp.Expr) == 0 {
+		return fmt.Errorf("promQL expression is not specified")
+	}
+	*d = tmp
+	return nil
+}
+
+func (d *lspData) returnPosition() (protocol.Position, error) {
+	if d.PositionLine == nil {
+		return protocol.Position{}, errors.New("positionLine is not specified")
+	}
+	if d.PositionChar == nil {
+		return protocol.Position{}, errors.New("positionChar is not specified")
+	}
+	return protocol.Position{
+		Line:      *d.PositionLine,
+		Character: *d.PositionChar,
+	}, nil
+}
+
+type API struct {
+	langServer    langserver.HeadlessServer
+	mdws          []middlewareFunc
+	enableMetrics bool
+}
+
+// NewLangServerAPI create a new instance of the Stateless API to use the LangServer through HTTP.
+//
+// If metadata is fetched from a remote Prometheus, the metadataService
+// implementation from the promql-langserver/prometheus package can be used,
+// otherwise you need to provide your own implementation of the interface.
+//
+// The provided Logger should be synchronized.
+//
+// In case "enableMetrics" is set to true, endpoint /metrics is then available and a middleware that instrument the different endpoints provided is instantiated.
```

Comment:
```suggestion
// In case "enableMetrics" is set to true, endpoint /metrics is then available and a middleware that instruments the different endpoints provided is instantiated.
```